workload/schemachange: avoid JSON columns in indexes for mixed version

fqazi opened the pull request:
Previously, we could generate CREATE TABLE / INDEX statements with JSON index columns which could end up failing in a mixed version state. To address this, this patch detects JSON columns and expects the appropriate errors in a mixed version state.

Fixes: #103875
Release note: None

fqazi committed:
workload/schemachange: avoid JSON columns in indexes for mixed version

Previously, we could generate CREATE TABLE / INDEX statements
with JSON index columns which could end up failing in
a mixed version state. To address this, this patch
detects JSON columns and expects the appropriate errors
in a mixed version state.

Fixes: #103875
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103973)
<!-- Reviewable:end -->

fqazi committed:
sql: add version gate for unique indexes of JSON types

Previously, when creating a table in a mixed version
state we incorrectly allowed unique indexes with JSON
types. This could cause errors on older nodes, which
do not know how to index this type. To address this,
this patch extends the version gate to unique indexes
during CREATE TABLE.

Epic: none

Release note: None

chengxiong-ruan commented on pkg/sql/create_table.go:
```diff
@@ -1603,7 +1603,7 @@ func NewTableDesc(
 			}
 
 			// Version gates for enabling primary keys for JSONB columns
```
nit: This comment would appreciate an update.

fqazi commented:
@chengxiong-ruan TFTR!

fqazi commented:
Hit a flaky test, retrying

bors r+

fqazi commented:
Hit an unrelated test flake, retrying one more time

bors r+
